src/Syntax/TimeAgnostic.hs

```diff
@@ -489,8 +476,15 @@ eval e = case e of
   ByStr _ s     -> pure . fromString $ s
   ByLit _ s     -> pure s
 
-  Eq  _ a b     -> [a' == b' | a' <- eval a, b' <- eval b]
-  NEq _ a b     -> [a' /= b' | a' <- eval a, b' <- eval b]
+  -- TODO better way to write these?
+  Eq _ SInteger x y -> [ x' == y' | x' <- eval x, y' <- eval y]
```

Comment:
Out of curiosity, why doesn't the following work?

```suggestion
  Eq _ _ x y -> [ x' == y' | x' <- eval x, y' <- eval y]
```

Reply:
It is not able to figure out that the type of `x'` and `y'` (which is `TypeOf a`)  have `Eq` instances. But once we make `a` concrete (e.g., `AInteger`) it knows that `TypeOf AInteger` is `Integer` and equality works.

Here, the singleton `SType` tag of `Eq`/`NEq` works as run-time witnesses of the `AType` parameter. So if we pattern match on `SInteger` we know that the `a` parameter of the expressions is `AInteger`.   I figured this while writing this code, and I though it was a very cool use of singleton types.